### DISCUSSION TITLE
Pattern for async init/shutdown for server modules

### DISCUSSION THREAD

**Discussion by humphd** [remix-run/remix#5745] [Q&A]
I'm looking for advice on the best pattern to use in order to initialize some server-side code that requires **async** logic, and how to do it so that it works well in dev (live reload) or prod.  For example, imagine a module `app/my-module.server.ts`, which needs to have an `init()` called in order to `fetch()` some file on startup, prior to being used.

This has been discussed in https://github.com/remix-run/remix/discussions/2949, but the various ways I see people doing this is usually for synchronous code paths.  For example, in the [ioredis example](https://github.com/remix-run/examples/blob/main/ioredis/app/utils/redis.server.ts):

```ts
import type { Redis as RedisType } from "ioredis";
import Redis from "ioredis";

let redis: RedisType;

declare global {
  var __redis: RedisType | undefined;
}

// this is needed because in development we don't want to restart
// the server with every change, but we want to make sure we don't
// create a new connection to the Redis with every change either.
if (process.env.NODE_ENV === "production") {
  redis = new Redis(process.env.REDIS_URL);
} else {
  if (!global.__redis) {
    global.__redis = new Redis(process.env.REDIS_URL);
  }
  redis = global.__redis;
}

export { redis };
```

I'm building something based on the [Blue Stack](https://github.com/remix-run/blues-stack), and my first attempt was to put an `init()` in [`server.ts` prior to requiring the server bundle](https://github.com/remix-run/blues-stack/blob/main/server.ts#L98),  However, while this works in dev, it fails in production builds.

Next, I thought maybe I could leverage `entry.server.ts`, but it doesn't provide any hook I can use.

Short of forcing something in, I'm wondering if there is a clean solution I'm overlooking?  I guess I could cache a `Promise` on the `global` in dev, and always re-`await` its value at runtime.  But it would be nice if I could have the app wait until I'm ready.

In a similar vein, what does async, graceful shutdown logic look like in `*.server.ts` modules?  For example, when you have to wait on some worker process to complete?

Thanks for any guidance on ways you've solved this.

**Comment by janhesters** [comment]
You can defer the promise and then `await` it in your loaders. Once the promised resolved the first time, it would be pretty much instant for each subsequent loader.

```ts
// mySDK.server.ts
export const mySDK = init(); // this is async and init returns a promise.

// Then in your loader, e.g. in app/routes/login.tsx
import { mySDK } from '~/mySDK.server';

export const loader = async () => {
  const sdk = await mySDK;
  // your logic ...
}
```
